src/python/pants/engine/target.py

```diff
@@ -0,0 +1,193 @@
+        field.
+        """
+
+
+class AsyncField(Field, metaclass=ABCMeta):
+    """A field that needs the engine in order to be hydrated.
+
+    You should create a corresponding Result class and define a rule to go from this AsyncField to
+    the Result. For example:
+
+        class Sources(AsyncField):
+            alias: ClassVar = "sources"
+            raw_value: Optional[List[str]]
+
+        @dataclass(frozen=True)
+        class SourcesResult:
+            snapshot: Snapshot
+
+
+        @rule
+        def hydrate_sources(sources: Sources) -> SourcesResult:
+            sources.validate_pre_hydration()
+            result = await Get[Snapshot](PathGlobs(sources.raw_value))
+            sources.validate_post_hydration()
+            return SourcesResult(result)
+
+
+        def rules():
+            return [hydrate_sources]
+
+    Then, call sites can `await Get` if they need to hydrate the field:
+
+        sources = await Get[SourcesResult](Sources, my_tgt.get(Sources))
+    """
+
+    def validate_pre_hydration(self) -> None:
+        """Any validation that can be done on the original `raw_value`.
+
+        It is cheaper to do any possible validation here, rather than in `validate_post_hydration`,
+        because we can short-circuit if an invariant is violated before incurring the expense of
+        hydration.
+        """
+
+    def validate_post_hydration(self, result: Any) -> None:
+        """Any validation that must be done after hydration by inspecting the result of that
+        hydration.
+
+        For example, a `PythonSources` field may validate that all hydrated source files end in
+        `.py`.
+        """
+
+
+class PluginField:
+    """Allows plugin authors to add new fields to pre-existing target types via UnionRules.
+
+    When defining a Target, authors should create a corresponding PluginField class marked with
+    `@union`. Then, plugin authors simply need to create whatever new `Field` they want and in a
+    `register.py`'s `rules()` function, call `UnionRule`. For example, to add a
+    `TypeChecked` field to `python_library`, register `UnionRule(PythonLibraryField, TypeChecked)`.
+
+        @union
+        class PythonLibraryField(PluginField):
+            pass
+
+
+        class PythonLibrary(Target):
+            core_fields = (Compatibility, PythonSources, ...)
+            plugin_field_type = PythonLibraryField
+
+
+        class TypeChecked(PrimitiveField):
+            ...
+
+
+        def rules():
+            return [UnionRule(PythonLibraryField, TypeChecked)]
+    """
+
+
+_F = TypeVar("_F", bound=Field)
+
+
+@frozen_after_init
+@dataclass(unsafe_hash=True)
+class Target(ABC):
+    # Subclasses must define these
+    alias: ClassVar[str]
+    core_fields: ClassVar[Tuple[Type[Field], ...]]
+    plugin_field_type: ClassVar[Type[PluginField]]
+
+    # These get calculated in the constructor
+    plugin_fields: Tuple[Type[Field], ...]
+    field_values: Dict[Type[Field], Any]
+
```

Comment:
Likely, we'll want to add an `Address` property. I believe every `Target` has an `Address`. I saved it for a followup because this PR is only focused on adding the core abstractions.